execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -252,16 +254,45 @@ where
         if !visited.insert(trie_key) {
             continue;
         }
-        let maybe_retrieved_trie: Option<Trie<K, V>> = store.get(txn, &trie_key)?;
-        match maybe_retrieved_trie {
-            // If we can't find the trie_key; it is missing and we'll return it
+        let maybe_retrieved_trie_bytes = store.get_raw(txn, &trie_key)?;
+        // Optimization: Don't look for descendants of leaves.
+        match maybe_retrieved_trie_bytes.and_then(|bytes| bytes.get(0)) {
+            // if this is a leaf don't parse (ie, the first byte is 0), just continue
+            Some(&Trie::<K, V>::LEAF_TAG) => continue,
+            // Either no entry under trie key or data is empty.  If data is empty then the trie is
+            // corrupted. Either way treat as missing.
             None => {
                 missing_descendants.push(trie_key);
+                continue;
+            }
+            _ => (),
+        }
+
+        // Parse the trie, handling errors gracefully.
+        let retrieved_trie: Trie<K, V> = match maybe_retrieved_trie_bytes
+            .map(bytesrepr::deserialize_from_slice)
+            .transpose()
+        {
+            Ok(Some(retrieved_trie)) => retrieved_trie,
+            Ok(None) => {
+                missing_descendants.push(trie_key);
+                continue;
+            }
+            // Couldn't parse; treat as missing and continue
+            Err(err) => {
+                error!(?err, "unable to parse trie");
+                missing_descendants.push(trie_key);
```

Comment:
I wonder whether we should delete the invalid data from the store now.
(On the other hand, maybe that's dangerous?)

Reply:
I'd rather leave the logic as is, since with this approach, the error gets logged and we also try to recover by marking the invalid data as missing and hence we try to fetch it from a peer.

I'm not sure it'd be dangerous as such to delete the corrupted data, but it would be a slightly bigger change as we'd need to change the method to accept a rw transaction, and this is not a pattern we generally use when encountering parsing errors/corrupted data.

In the unoptimised code currently in `dev`, I believe we actually just exit early, returning an error in this case.